riscv/src/elf.rs

```diff
@@ -0,0 +1,1213 @@
+                    }
+                    // s{b|h|w} rd, symbol, rt
+                    Ins {
+                        opc: l_op,
+                        rd: None,
+                        rs1: Some(rt_l),
+                        rs2: Some(_),
+                        imm: Some(lo),
+                        ..
+                    } if matches!(l_op, Op::SB | Op::SH | Op::SW) && rd_auipc == rt_l => {
+                        // We don't support code modification, so it is better
+                        // to panic if this is the case:
+                        let addr = hi.wrapping_add(*lo);
+                        assert!(!self.address_map.is_in_text_section(addr as u32));
+
+                        // Otherwise, this is a data store instruction. To be
+                        // more conformant, it is better to let two
+                        // instructions be handled separately.
+                        return None;
+                    }
+                    // call offset
+                    Ins {
+                        opc: Op::JALR,
+                        rd: Some(1),
+                        rs1: Some(1),
+                        rs2: None,
+                        imm: Some(lo),
+                        ..
+                    } if *rd_auipc == 1 => HighLevelInsn {
+                        op: "call",
+                        args: HighLevelArgs {
+                            imm: HighLevelImmediate::CodeLabel(hi.wrapping_add(*lo) as u32),
+                            ..Default::default()
+                        },
+                        original_address,
+                    },
+                    // tail offset
+                    Ins {
+                        opc: Op::JALR,
+                        rd: Some(0),
+                        rs1: Some(6),
+                        rs2: None,
+                        imm: Some(lo),
+                        ..
+                    } if *rd_auipc == 6 => HighLevelInsn {
+                        op: "tail",
+                        args: HighLevelArgs {
+                            imm: HighLevelImmediate::CodeLabel(hi.wrapping_add(*lo) as u32),
+                            ..Default::default()
+                        },
+                        original_address,
+                    },
+                    _ => {
+                        panic!(
+                            "Unexpected instruction after AUIPC: {insn2:?} at {original_address:08x}"
+                        );
+                    }
+                }
+            }
+            _ => return None,
+        };
+
+        // TODO: implement here other kinds of RISC-V fusions as optimization.
+
+        if let (ReadOrWrite::Write(refs), HighLevelImmediate::CodeLabel(addr)) =
+            (&mut self.referenced_text_addrs, &result.args.imm)
+        {
+            refs.insert(*addr);
+        }
+
+        Some(result)
+    }
+
+    fn map_one(&mut self, insn: MaybeInstruction) -> HighLevelInsn {
```

Comment:
Map one what? Or is it the literal 1?

Reply:
This is a trait implementation. See trait `TwoOrOneMapper`.